src/proto/device_path/text.rs

```diff
@@ -0,0 +1,122 @@
+//! `DevicePathToText` and `DevicePathFromText` Protocol
+
+use crate::{
+    proto::{device_path::DevicePath, Protocol},
+    unsafe_guid, CStr16, Char16,
+};
+
+/// Device Path to Text protocol.
+///
+/// This protocol provides common utility functions for converting device
+/// nodes and device paths to a text representation.
+#[repr(C)]
+#[unsafe_guid("8b843e20-8132-4852-90cc-551a4e4a7f1c")]
+#[derive(Protocol)]
+pub struct DevicePathToText {
+    convert_device_node_to_text: unsafe extern "efiapi" fn(
+        device_node: *const DevicePath,
+        display_only: bool,
+        allow_shortcuts: bool,
+    ) -> *const Char16,
+    convert_device_path_to_text: unsafe extern "efiapi" fn(
+        device_path: *const DevicePath,
+        display_only: bool,
+        allow_shortcuts: bool,
+    ) -> *const Char16,
+}
+
+impl DevicePathToText {
+    /// Convert a device node to its text representation.
+    ///
+    /// The `display_only` parameter controls whether the longer (parseable) or
+    /// shorter (display-only) form of the conversion is used. If `display_only`
+    /// is TRUE, then the shorter text representation of the display node is
+    /// used, where applicable. If `display_only` is FALSE, then the longer text
+    /// representation of the display node is used.
+    ///
+    /// The `allow_shortcuts` is FALSE, then the shortcut forms of text
+    /// representation for a device node cannot be used. A shortcut form
+    /// is one which uses information other than the type or subtype. If
+    /// `allow_shortcuts is TRUE, then the shortcut forms of text
+    /// representation for a device node can be used, where applicable.
+    ///
+    /// Returns `None` if `device_node` was NULL or there was insufficient memory.
+    pub fn convert_device_node_to_text(
+        &self,
+        device_node: &DevicePath,
+        display_only: bool,
+        allow_shortcuts: bool,
+    ) -> Option<&CStr16> {
```

Comment:
These functions allocate from UEFI boot services memory. Nothing here frees the memory, so it will leak. We have solved this in other places by wrapping the returned pointer in a struct with a custom `Drop` implementation. See [`ProtocolsPerHandle`](https://github.com/rust-osdev/uefi-rs/blob/be4f39cfc37901c93ce6381994460b5fa8f574a1/src/table/boot.rs#L1559) for example. It would probably be good to make a shared extraction for this pattern, but doesn't necessarily need to be done in this PR.

Reply:
Should I do some changes for this?

Reply:
I had to think a bit on this :) Here's what I would like added here:

```rust
/// Wrapper for a string internally allocated from UEFI boot services memory.
pub struct PoolString<'a> {
    boot_services: &'a BootServices,
    text: *mut Char16,
}

impl<'a> Drop for PoolString<'a> {
    fn drop(&mut self) {
        let _ = self.boot_services.free_pool(self.text as *mut u8);
    }
}

impl<'a> Deref for PoolString<'a> {
    type Target = CStr16;

    fn deref(&self) -> &CStr16 {
        unsafe { CStr16::from_ptr(self.text) }
    }
}

impl<'a> PoolString<'a> {
    fn new(boot_services: &'a BootServices, text: *mut Char16) -> Option<Self> {
        if text.is_null() {
            None
        } else {
            Some(Self {
                boot_services,
                text,
            })
        }
    }
}
```

Then we can use it like so:

```rust
pub fn convert_device_node_to_text<'boot>(
    &self,
    boot_services: &'boot BootServices,
    device_node: &DevicePath,
    display_only: DisplayOnly,
    allow_shortcuts: AllowShortcuts,
) -> Option<PoolString<'boot>> {
    unsafe {
        PoolString::new(
            boot_services,
            (self.convert_device_node_to_text)(device_node, display_only.0, allow_shortcuts.0),
        )
    }
}
```

Note the new `boot_services` parameter, and associated lifetime. And same change should be made to `convert_device_path_to_text` of course.

This also requires changing the `extern fn`s so that they return a `*mut Char16` pointer rather than `*const`, as the `free_pool` function expects a mutable pointer.

Reply:
Thanks a lot, it's very helpful!😃

Reply:
Thanks a lot, it's very helpful!😃